[Docs] finishing account crate in dpc (#4)
lambdaclass/aleo_internals_book: src/snarkVM/dpc/ledger.md

```rust 
/// Adds the given unconfirmed transaction to the memory pool.
pub fn add_unconfirmed_transaction(&mut self, transaction: &Transaction<N>) -> Result<()> {
    ensure!(
        self.canon_blocks.contains_ledger_root(&transaction.ledger_root()), 
        "Transaction references a non-existent ledger root"
    );

    // Ensure the transaction does not contain serial numbers already in the canon chain.
    for serial_number in transaction.serial_numbers() {
        ensure!(
            !self.canon_blocks.contains_serial_number(serial_number), 
            "Transaction contains a serial number already in existence"
        );
    }

    // Ensure the transaction does not contain commitments already in the canon chain.
    for commitment in transaction.commitments() {
        ensure!(
            !self.canon_blocks.contains_commitment(commitment), 
            "Transaction contains a commitment already in existence"
        );
    }

    // Attempt to add the transaction into the memory pool.
    self.memory_pool.add_transaction(transaction)?;

    Ok(())
}
```

> A transaction needs to reference the canonical chain's root, and all of its serial numbers and it's commintment need to be unused in the canonical chain to be included in the transaction memory pool.

Reviewer: ```suggestion
A transaction needs to reference the canonical chain's root, and all of its serial numbers and its commitment need to be unused in the canonical chain to be included in the transaction memory pool.
```

Author: This change was resolved on the parent PR that has been merged already #1.

The juciest method is the `mine_next_block` method which has a few steps:
 

> - Ensure that the new timestamp is ahead of the previous timestamp (it's more of a fix it if it is not ahead, it won't throw an error if it is behind).

Reviewer: ```suggestion
- Ensure that the new timestamp is ahead of the previous timestamp (it's more of a fix if it is not ahead, it won't throw an error if it is behind).
```

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)

    current_hash: N::BlockHash,
    /// The current ledger tree.
    ledger_tree: LedgerTree<N>,
    /// The chain of previous block hashes.
    previous_hashes: HashMap<u32, N::BlockHash>,
    /// The chain of block headers.
    headers: HashMap<u32, BlockHeader<N>>,
    /// The chain of block transactions.
    transactions: HashMap<u32, Transactions<N>>,
}
```
Most methods in `blocks.rs` are getters/setters that delegate the responsibility to either the `ledger_tree`, `previous_hashes`, `headers`, or `transactions`.

One of the 'heavy' functions is the `add_next` method, which adds a block as the latest in the canonical chain and is used by the `Ledger`. 
It runs some verifications by ensuring:
- The given block is valid
- The given block has the correct block height
- The given block height does not already exist in the ledger
- The given block has the appropiate previous block hash
- The given block hash does not already exist in the ledger
- The next block timestamp is within the declared time limit
- The given block timestamp is after the current timestamp
- The given block difficulty target is correct
- The expected cumulative weight is computed correctly

> - The given cumulative weight is correct.

Reviewer: ```suggestion
- The given cumulative weight is correct
```

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)

- Each transaction in the block does not already exist in the ledger
- Each transaction's ledger root is the same as the ledger's
- Each serial number does not exist in the ledger

> - Each commitment does not exist in the ledger

Reviewer: ```suggestion
- Each commitment does not exist in the ledger.
```

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)

And after that it attempts to add the block in each structure.

```rust 
// Add the block to the ledger. This code section executes atomically.
{
    let mut blocks = self.clone();

    blocks.current_height = height;
    blocks.current_hash = block_hash;
    blocks.ledger_tree.add(&block.hash())?;
    blocks.previous_hashes.insert(height, block.previous_block_hash());
    blocks.headers.insert(height, block.header().clone());
    blocks.transactions.insert(height, block.transactions().clone());

    *self = blocks;
}
```

**question**: why does this execute atomically? Did they actually meant that since the self is cloned you don't have race conditions? because the block does not execute atomically. 
    Is it possible to add two different blocks with the same height twice? -> No, since each `add_next` call overwrites the whole structure
    

> The next method to take a look at is the to_ledger_proof. This does a bunch of things but succintly, it computes a few other proofs:

Reviewer: ```suggestion
The next method to take a look at is the to_ledger_proof. This does a bunch of things but succinctly, it computes a few other proofs:
```

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)

```mermaid
flowchart LR

a[LedgerProof] --> b[LedgerRootInclusionProof]
a[LedgerProof] --> c[RecordProof]

c[RecordProof] --> d[BlockHeaderInclusionProof]
c[RecordProof] --> e[TransactionsInclusionProof]
c[RecordProof] --> f[LocalInclusionProof]

d[BlockHeaderInclusionProof] --> g[BlockHeader::to_header_inclusion_proof]
e[TransactionsInclusionProof] --> h[Transactions::to_transactions_inclusion_proof]
f[LocalInclusionProof] --> i[Transitions::to_local_proof]
b[LedgerRootInclusionProof] --> j[LedgerTree::to_ledger_inclusion_proof]

> i[Transitions::to_local_proof] --> k[Transition::to_transition_inclusion_proof]

Reviewer: ```suggestion
i[Transitions::to_local_proof] --> k[**Transition**::to_transition_inclusion_proof]
```

Is it ok for "Transition" to be singular? All the other ones are in plural.

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)

///     T_{i+1} = T_anchor * 2^((S - B * N) / tau).
///     T_anchor = Anchor target of a specific block height
///     B = Expected time per block.
///     S = Time elapsed since the anchor.
///     N = Number of blocks since the anchor.
///     tau = The half life of the algorithm. For every `tau` seconds ahead of
///           schedule a block’s timestamp becomes, the difficulty doubles.
/// To avoid use of floating points, we use fixed-point arithmetic.
fn asert_retarget(
    anchor_timestamp: i64,
    anchor_difficulty_target: u64,
    anchor_block_height: u32,
    block_timestamp: i64,
    block_height: u32,
    target_block_time: i64,
) -> u64 ```

## LedgerTree
The ledger tree is a merkle tree
```rust
pub struct LedgerTree<N: Network> {
    #[derivative(Debug = "ignore")]
    tree: Arc<MerkleTree<N::LedgerRootParameters>>,
    block_hashes: HashMap<N::BlockHash, u32>,
    current_index: u32,
}
```

The `tree` is the actual Merkle Tree, but the `block_hashes` are used to keep track of the blocks in the tree (lookup is faster) and to map the block hash to its respective height in the tree. The `current index keeps track of the index of the last added block in the tree`

> There are two methods for adding blocks, `add` and `add_all`. Those are pretty much well documented. One thing i noticed is this piece:

Reviewer: ```suggestion
There are two methods for adding blocks, `add` and `add_all`. Those are pretty much well documented. One thing I noticed is this piece:
```

Author: Already resolved on [f3642ee](https://github.com/lambdaclass/aleo_internals_book/pull/4/commits/f3642ee82c142948e3c2880446ed1c89044b9a43)
